### DISCUSSION TITLE
SPI trouble - MCP2515

### DISCUSSION THREAD

**Discussion by treyus30** [earlephilhower/arduino-pico#1536] [Q&A]
I have confirmed my SPI pins are working through an Adafruit display library. 

When I repurpose the same pins with [this library](https://github.com/autowp/arduino-mcp2515), the device softlocks in the same way as if it has improperly assigned pins once uploaded. I've spent several hours trying to debug and I'm ready to seek help. The library itself is not that complex and it works flawlessly on a standard Arduino Nano. 

Basic test code (SS can be replaced with csPin):
```
#include <SPI.h>
#include <mcp2515.h>
 

int cntr = 0;
unsigned long oldTime = 0;
 
int sckPin = 2;
int mosiPin = 3;
int misoPin = 0;
int csPin = 1;

struct can_frame canMsg;
  MCP2515 mcp2515(SS);

void setup() {
  Serial.begin(9600);
  SPI.setCS(csPin);
  SPI.setRX(misoPin);
  SPI.setTX(mosiPin);
  SPI.setSCK(sckPin);
  pinMode(25, OUTPUT);
  
  mcp2515.reset();
  mcp2515.setBitrate(CAN_1000KBPS, MCP_8MHZ);
  mcp2515.setNormalMode();
 
  Serial.println("------- CAN Speedtest ----------");
}
 
void loop() {
  digitalWrite(25, HIGH); // Backlight on
  delay(500);
  digitalWrite(25, LOW); // Backlight off
  delay(500);

  if (mcp2515.readMessage(&canMsg) == MCP2515::ERROR_OK) {
    cntr++;
  }
 
  if ((millis()-oldTime)>1000) {
    oldTime = millis();
    Serial.print(cntr);
    Serial.println(" msg/sec");
    cntr = 0;      
  }
}
```

**Comment by earlephilhower** [comment]
The library is performing illegal access to unitialized global variables (SPI in this case) when you use a global `MCP2525`.
https://github.com/autowp/arduino-mcp2515/blob/d9adb28ca34b6a22677bb297a9c862015ba4d057/mcp2515.cpp#L15-L30

Make the MCP2525 a `MCP2515 *mcp2515;` and then do a `mcp2515 = new MCP2515(SS)` at the top of your setup and then use pointer access elsewhere and you should be OK.

In general there is no order guarantee when globals are initted.  In this instance, the MCP global is being initted before the SPI ones.  But, it could be the other way, depending on the compiler, the linker, other bits of code, phase of the moon, etc...
